AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/RemoteSyncEngine+Resolver.swift

```diff
@@ -23,9 +23,20 @@ extension RemoteSyncEngine {
             case (.pausingMutationQueue, .pausedMutationQueue(let storageEngineAdapter)):
                 return .clearingStateOutgoingMutations(storageEngineAdapter)
 
-            case (.clearingStateOutgoingMutations, .clearedStateOutgoingMutations(let api, let storageEngineAdapter)):
+            case (.clearingStateOutgoingMutations, .clearedStateOutgoingMutations):
+                return .checkIfAuthenticationRequired
+            case (.checkIfAuthenticationRequired, .requireAuthenticatedUser):
+                return .waitForAuthenticatedUser
+            case (.waitForAuthenticatedUser, .authenticatedUser(let api,
+                                                                let storageEngineAdapter,
+                                                                let userId)):
+                return .initializingSubscriptions(api, storageEngineAdapter, userId)
+            case (.checkIfAuthenticationRequired, .authenticatedUser(let api,
+                                                                     let storageEngineAdapter,
+                                                                     let userId)):
+                return .initializingSubscriptions(api, storageEngineAdapter, userId)
+            case (.checkIfAuthenticationRequired, .authenticationNotRequired(let api, let storageEngineAdapter)):
```

Comment:
Seems like we've adopted a strategy where we handle listening to auth state within the sync engine.  Did we consider another approach like listening to auth state in StorageEngine and determine whether or not we should start/stop the sync engine from there?

Reply:
Been chatting with @elorzafe regarding the auth strategy and we agree that DataStore should listen to auth state changes in general 

JS current requires the developers to listen to auth state changes themselves:
@elorzafe, please correct me if i'm wrong here:
If sign-in event: no public facing API to start the sync engine, but can call `DataStore.clear()` which re-iniitializes the syncEngine as a secondary behavior
if sign out event: call DataStore.clear() to clear local data

JS customers have to make a call to DataStore APIs like DataStore.query(). this wll try to
- start to start the sync engine (fail if not authenticated, fall back to IAM credentials for sigv4 signing, different architecture from iOS intercategory plugin communication - IOS cannot fall back to IAM, but utimately there is an eventual fail case). 
- assuming this is async, the data returned to the customer may not be up to date since sync engine's reconcilation path may not have caught up.

The most efficient time to start the sync engine is by listening to auth sign in event. if all customers which use auth enabled models require the user to be signed in before making API calls (subscriptions and sync query) then DataStore should hold this listening logic. it should be waiting for Auth events like sign-in and sign-out.

The different places this auth state check can live:
1. At the very top level, DataStorePlugin will wait for DataStore configured event, check for API plugin before starting the sync engine. the check if there are required API optionally Auth plugin can be expanded to auth state change listener here
2. As you mentioned, in the StorageEngine, this looks like a possible candidate and be a good place when we address the auth state listener for sign-out because the `clear()` API is there. (we may then refactor the auth state lisnter code for sign-in event to the same place)
3. where it is currently in syncEngine: this allows the sync engine to do some of the non-auth related house keeping work that can be done without the API calls (subscription and sync query). leaving the wait for auth state change in the most optimal spot. and SyncEngine already exists a state machine like `schedulingRestart` and `errored` which will allow us to easily add "expoential retry if user is not signed in, then give up and listen for auth state change"

Reply:
- refactored code to put it in StorageEngine
- also added the signOut listener

Reply:
removed auth listener code (removed sign in/sign out listener code) for less impactful behavior change